File: hammerspoon/cr/trigger.lua
-- cr.trigger — the per-reminder state machine ("the guard").
--
-- The observer samples a LEVEL (what's on screen); this module detects EDGES
-- (the thing a reminder cares about just ended). Completion conditions are
-- edge-triggered: you must see the activity happening, then see it stop —
-- otherwise "not watching the video" is trivially true and fires instantly.
--
--   PENDING ──seen──▶ ARMED ──absent──▶ COOLDOWN ──absent × N──▶ READY ──gate──▶ FIRED
--                       ▲                  │ present                │ present
--                       └──────────────────┴──────── re-arm ────────┘
--
-- Gates for firing (never mid-focus): an app-switch/wake tick, or a max-wait
-- backstop so a reminder can't sit READY forever.

local config = require("cr.config")
local log = require("cr.log")
local matcher = require("cr.matcher")
local reminders = require("cr.reminders")
local notifier = require("cr.notifier")
local observer = require("cr.observer")
local watchfor = require("cr.watchfor")
local tier = require("cr.tier")

local M = { running = false }

-- wired by cr.init (e.g. menubar badge refresh); called after any state change
M.onStateChange = nil

local function cfg()
  return config.trigger
end

-- The card for a fired reminder. Built separately from fire() so an unanswered
-- reminder can be put back on screen without re-firing it.
-- Defer by a fixed number of minutes, from now. Works for both kinds: a timed
-- reminder moves its clock, a contextual one stops watching and comes back on
-- the clock instead — otherwise "in 5 minutes" would mean "in 5 minutes, if
-- you also happen to finish that thing", which is not what anyone means.
local function deferBy(r, minutes, label)
  r.dueAt = os.time() + minutes * 60
  reminders.setState(r, "scheduled", label)
  log.append({ event = "feedback", value = "defer", minutes = minutes, id = r.id })
  require("cr.why").note("reminder pushed back", r.text, {
    { "you pressed", label },
    { "now due", os.date("%I:%M %p", r.dueAt):gsub("^0", "") },
  })
end

-- The card for a fired reminder. Built separately from fire() so an unanswered
-- reminder can be put back on screen without re-firing it.
local function cardFor(r, timed)
  -- ids are stable, labels are not: "Too early" became "5 min" during this
  -- project, and anything that had addressed that action by label would have
  -- broken silently. See cr.notification.
  local actions = {
    {
      id = "done",
      label = "Done",
      fn = function()
        reminders.setState(r, "done", "user")
        log.append({ event = "feedback", value = "done", id = r.id })
      end,
    },
    {
      -- The common answer to a reminder is "yes, but not this second", and it
      -- was previously two clicks away behind a generic Snooze. Naming the
      -- interval means you don't have to remember what Snooze is set to.
      id = "defer_5",
      label = "5 min",
      fn = function() deferBy(r, 5, "5 min") end,
    },
    {
      id = "snooze",
      label = "Snooze",
      fn = function() deferBy(r, cfg().snoozeMinutes, "Snooze") end,
    },
  }
  return {
    title = "Reminder",
    body = r.text,
    icon = "⏰",
    urgency = "info",
    sticky = true, -- waits for an answer; see cr.notify_ui
    meta = { id = r.id, referent = r.referent and r.referent.label },
    actions = actions,
  }
end

-- Why this reminder is arriving at this moment, in the terms the user would
-- use. The two paths have genuinely different answers: a clock reminder fires
-- because a time arrived, a contextual one because it watched an activity end.
local function whyNow(r, snap, gate)
  if gate == "time" then
    local late = os.time() - (r.dueAt or os.time())
    return string.format("the time you set arrived (%s)%s",
      os.date("%I:%M %p", r.dueAt or os.time()):gsub("^0", ""),
      late > 2 and string.format(" — %ds later than set, checked once a second", late) or "")
  end
  local where = r.referent and r.referent.label or "?"
  if gate == "max-wait" then
    return string.format('you left "%s" and stayed away, and nothing interrupted for %ds — '
      .. "fired rather than wait longer", where, cfg().maxReadyWait)
  end
  return string.format('you left "%s" and stayed away for %d checks (~%ds), then %s — '
    .. "waited for that seam so this didn't land mid-task",
    where, cfg().absentSamples, cfg().absentSamples * (config.pollInterval or 5),
    gate == "wake" and "you came back from idle" or "you switched apps")
end

local function fire(r, snap, gate)
  reminders.setState(r, "fired", gate)
  log.append({ event = "trigger.fired", id = r.id, text = r.text, gate = gate,
               tier = r.tier })

  -- Ambient never gets a card. This is the whole point of the tier: something
  -- you asked to "keep an eye on" should be findable, not interruptive, and a
  -- card that must be dismissed is an interruption whatever its contents say.
  -- It still fires and still lands in the dashboard and menu bar.
  if (r.tier or tier.UPCOMING) == tier.AMBIENT then
    log.append({ event = "trigger.silent", id = r.id, reason = "ambient tier" })
    require("cr.why").note("reminder surfaced quietly", r.text, {
      { "why now", whyNow(r, snap, gate) },
      { "attention", "ambient — no card, by design. It's in the dashboard and menu bar." },
    })
    return
  end
  local payload = cardFor(r, gate == "time")
  local chans = r.channels
  if (r.tier or 2) == tier.CRITICAL then
    payload.urgency = "warn"
    payload.icon = "🚨"
    -- critical reaches you wherever you are, not only where it fired
    chans = { "card", "system" }
    for _, c in ipairs(r.channels or {}) do
      if c ~= "card" and c ~= "system" then chans[#chans + 1] = c end
    end
    local s_ = hs.sound.getByName("Sosumi"); if s_ then s_:play() end
  end
  local channels = notifier.notify(payload, {
    channels = chans,
    event = "reminder.fired",
    reminder = r,
    -- the same sentence the decision log gets. On a phone there is no screen
    -- context to reconstruct why this arrived, so the reason travels with it.
    trigger = { gate = gate, why = whyNow(r, snap, gate) },
  })
  require("cr.why").note("reminder fired", r.text, {
    { "why now", whyNow(r, snap, gate) },
    { "set", os.date("%I:%M %p", r.createdAt or os.time()):gsub("^0", "")
        .. (r.whenPhrase and (' from "' .. r.whenPhrase .. '"') or "") },
    { "delivered to", table.concat(channels or r.channels or { "card" }, " + ") },
    { "waiting on", "you — the card stays up until you answer it" },
  })
end

-- Feed one OCR capture to every reminder waiting on content, and fire the ones
-- whose condition just became true. Wired to screen_text.onCapture in cr.init.
--
-- Note what this does NOT do: it never fires on a first sighting. cr.watchfor
-- requires seeing the subject *pending* before it will accept *done*, which is
-- the same edge-triggered rule the window FSM above uses and for the same
-- reason — when you say "merge the PR after the tests run", last run's "12
-- passed" is usually still sitting in the terminal, and a level check would
-- fire instantly on stale output and look broken.
function M.onCapture(entry)
  if not entry or not entry.text then return end
  local waiting = {}
  for _, r in ipairs(reminders.active()) do
    if r.watchFor then waiting[#waiting + 1] = r end
  end
  if #waiting == 0 then return end

  local fired = watchfor.evaluate(entry.text, waiting, log)
  for _, hit in ipairs(fired) do
    local r = hit.reminder
    reminders.persist()   -- seenPending/doneAt were mutated in place
    fire(r, observer.current, "content")
    require("cr.why").note("reminder fired", r.text, {
      { "why now", string.format('you asked to be reminded %s, and the screen said so: "%s"',
          r.watchFor.phrase or "when it finished", hit.evidence or "?") },
      { "how", "watched for it to be running first, then finish — so old output "
          .. "already on screen couldn't set it off" },
    })
  end
  if #fired > 0 and M.onStateChange then pcall(M.onStateChange) end
end

-- A sticky card only survives as long as the process drawing it. Hammerspoon
-- reloads (and reboots), and the reminder underneath stays "fired" with
-- nothing on screen — dismissed by accident rather than by the user. Put those
-- back up on start. Local card only: the remote channels already delivered
-- once, and re-pinging Discord on every reload would be its own bug.
function M.restoreFired()
  local n = 0
  for _, r in ipairs(reminders.active()) do
    if r.state == "fired" then
      n = n + 1
      notifier.notify(cardFor(r, r.dueAt ~= nil), {
        channels = { "card" },
        noMirror = true, -- a redraw, not a new event; must not re-ping remotes
        event = "reminder.restored",
        reminder = r,
      })
    end
  end
  if n > 0 then log.append({ event = "trigger.restored", count = n }) end
end

-- Tier is a property of the reminder *right now*, so it is recomputed rather
-- than remembered. "Take out the trash" is ambient in the afternoon, upcoming
-- in the evening, and critical the night before collection — nothing about the
-- reminder changed except how close the consequence got.
local function retier(r, snap)
  local from = r.tier or tier.UPCOMING
  local to, why = from, nil

  if r.dueAt then
    local left = r.dueAt - os.time()
    if left <= 600 and from > tier.CRITICAL then
      to, why = tier.CRITICAL, "under ten minutes away"
    elseif left <= 3600 and from > tier.UPCOMING then
      to, why = tier.UPCOMING, "within the hour"
    end
  elseif r.referent and from == tier.AMBIENT and matcher.matches(r.referent, snap) then
    -- something you were only tracking is now in front of you
    to, why = tier.INCONTEXT, "the thing it's about is on screen now"
  elseif from == tier.INCONTEXT and r.state == "pending"
      and os.time() - (r.createdAt or os.time()) > 86400 then
    -- a context that hasn't come back in a day isn't going to interrupt well
    to, why = tier.AMBIENT, "its context hasn't appeared in a day"
  end

  if to ~= from then
    r.tier, r.tierWhy = to, why
    r.tierHistory = r.tierHistory or {}
    r.tierHistory[#r.tierHistory + 1] = { t = os.time(), from = from, to = to, why = why }
    reminders.persist()
    log.append({ event = "tier.moved", id = r.id, from = from, to = to, why = why })
    require("cr.why").note(
      to < from and "reminder got louder" or "reminder got quieter", r.text, {
        { "moved", string.format("%s → %s", tier.LABEL[from].name, tier.LABEL[to].name) },
        { "because", why },
        { "now", tier.LABEL[to].note },
      })
    if M.onStateChange then pcall(M.onStateChange) end
  end
end

-- Timed reminders bypass the FSM entirely: they fire on the clock, wherever
-- the user is. Checked every second for demo-friendly precision.
local function checkDue()
  local now = os.time()
  for _, r in ipairs(reminders.active()) do
    if r.state == "scheduled" and r.dueAt and now >= r.dueAt then
      local ok, err = pcall(fire, r, observer.current or {}, "time")
      if not ok then
        log.append({ event = "trigger.error", id = r.id, error = tostring(err) })
      end
      if M.onStateChange then pcall(M.onStateChange) end
    end
  end
end

local function step(r, snap)
  local present = matcher.matches(r.referent, snap)

  if r.state == "pending" then
    if present then
      r.absent = 0
      reminders.setState(r, "armed")
    end

  elseif r.state == "armed" then
    if not present then
      r.absent = 1
      reminders.setState(r, "cooldown")
    end

  elseif r.state == "cooldown" then
    if present then
      -- brief tab-away, not "done" — the debounce doing its job
      r.absent = 0
      reminders.setState(r, "armed", "rearmed")
    else
      r.absent = (r.absent or 0) + 1
      if r.absent >= cfg().absentSamples then
        r.readyAt = os.time()
        reminders.setState(r, "ready")
      end
    end

  elseif r.state == "ready" then
    if present then
      -- came back before we fired: definitely not done
      r.absent = 0
      reminders.setState(r, "armed", "returned")
    else
      local gate
      if tier.bypassesSeamGate(r.tier) then
        -- The one tier allowed to interrupt mid-focus. Everything else waiting
        -- for a natural break is the promise that makes this tolerable to
        -- leave running; critical is the stated exception, not an oversight.
        gate = "critical"
      elseif snap.reason == "app-switch" or snap.reason == "wake" then
        gate = snap.reason
      elseif os.time() - (r.readyAt or 0) >= cfg().maxReadyWait then
        gate = "max-wait"
      end
      if gate then fire(r, snap, gate) end
    end

  elseif r.state == "snoozed" then
    if os.time() >= (r.snoozeUntil or 0) then
      r.readyAt = os.time()
      reminders.setState(r, "ready", "snooze-elapsed")
    end
  end
end

-- Exported for scripted tests and the future replay harness: feed synthetic
-- snapshots through the FSM without waiting for wall-clock time.
function M.tick(snap)
  for _, r in ipairs(reminders.active()) do
    pcall(retier, r, snap)
    local before = r.state
    local ok, err = pcall(step, r, snap)
    if not ok then
      log.append({ event = "trigger.error", id = r.id, error = tostring(err) })
    end
    if r.state ~= before and M.onStateChange then pcall(M.onStateChange) end
  end
end

local dueTimer

function M.start()
  if M.running then return end
  M.running = true
  observer.subscribeTick(M.tick)
  dueTimer = hs.timer.doEvery(1, checkDue)
  log.append({ event = "trigger.start" })
end

return M

File: windows/crw/trigger.py
"""The per-reminder state machine ("the guard") — port of cr/trigger.lua.

The observer samples a LEVEL (what's on screen); this module detects EDGES
(the thing a reminder cares about just ended). Completion is edge-triggered:
you must see the activity happening, then see it stop — otherwise "not
watching the video" is trivially true and fires instantly.

  PENDING ──seen──▶ ARMED ──absent──▶ COOLDOWN ──absent × N──▶ READY ──gate──▶ FIRED
                      ▲                  │ present                │ present
                      └──────────────────┴──────── re-arm ────────┘

Gates for firing (never mid-focus): an app-switch/wake tick, or a max-wait
backstop so a reminder can't sit READY forever.
"""

from __future__ import annotations

import time

from . import config, eventlog, matcher, notifier, reminders, tier, why

# wired by the app (glance refresh etc.); called after any state change
on_state_change = None


def _notify_change():
    if on_state_change:
        try:
            on_state_change()
        except Exception:
            pass


def defer_by(r, minutes, label):
    """Works for both kinds: a timed reminder moves its clock, a contextual
    one stops watching and comes back on the clock instead — otherwise "in 5
    minutes" would mean "in 5 minutes, if you also happen to finish that
    thing", which is not what anyone means."""
    r["dueAt"] = int(time.time()) + minutes * 60
    reminders.set_state(r, "scheduled", label)
    eventlog.append({"event": "feedback", "value": "defer",
                     "minutes": minutes, "id": r["id"]})
    why.note("reminder pushed back", r["text"], [
        ("you pressed", label),
        ("now due", time.strftime("%I:%M %p", time.localtime(r["dueAt"])).lstrip("0")),
    ])


def card_for(r):
    """The card for a fired reminder. Built separately from fire() so an
    unanswered reminder can be put back on screen without re-firing it.
    Action ids are stable, labels are not — see cr.notification."""
    def done():
        reminders.set_state(r, "done", "user")
        eventlog.append({"event": "feedback", "value": "done", "id": r["id"]})
        _notify_change()

    return {
        "title": "Reminder",
        "body": r["text"],
        "icon": "⏰",
        "urgency": "info",
        "sticky": True,  # waits for an answer; a card that dismisses itself
                         # has failed at its only job, invisibly
        "meta": {"id": r["id"],
                 "referent": (r.get("referent") or {}).get("label")},
        "actions": [
            {"id": "done", "label": "Done", "fn": done},
            {"id": "defer_5", "label": "5 min",
             "fn": lambda: (defer_by(r, 5, "5 min"), _notify_change())},
            {"id": "snooze", "label": "Snooze",
             "fn": lambda: (defer_by(r, config.SNOOZE_MINUTES, "Snooze"),
                            _notify_change())},
        ],
    }


def why_now(r, snap, gate) -> str:
    """Why this reminder is arriving at this moment, in the user's terms."""
    if gate == "time":
        due = r.get("dueAt") or int(time.time())
        late = int(time.time()) - due
        clock = time.strftime("%I:%M %p", time.localtime(due)).lstrip("0")
        extra = (f" — {late}s later than set, checked once a second"
                 if late > 2 else "")
        return f"the time you set arrived ({clock}){extra}"
    where = (r.get("referent") or {}).get("label") or "?"
    if gate == "max-wait":
        return (f'you left "{where}" and stayed away, and nothing interrupted '
                f"for {config.MAX_READY_WAIT}s — fired rather than wait longer")
    how = ("you came back from idle" if gate == "wake" else "you switched apps")
    return (f'you left "{where}" and stayed away for {config.ABSENT_SAMPLES} '
            f"checks (~{config.ABSENT_SAMPLES * config.POLL_INTERVAL}s), then "
            f"{how} — waited for that seam so this didn't land mid-task")


def fire(r, snap, gate) -> None:
    reminders.set_state(r, "fired", gate)
    eventlog.append({"event": "trigger.fired", "id": r["id"], "text": r["text"],
                     "gate": gate, "tier": r.get("tier")})

    # Ambient never gets a card: something you asked to "keep an eye on"
    # should be findable, not interruptive. It still lands in the dashboard
    # and the glance panel.
    if (r.get("tier") or tier.UPCOMING) == tier.AMBIENT:
        eventlog.append({"event": "trigger.silent", "id": r["id"],
                         "reason": "ambient tier"})
        why.note("reminder surfaced quietly", r["text"], [
            ("why now", why_now(r, snap, gate)),
            ("attention", "ambient — no card, by design. It's in the dashboard "
                          "and the glance panel."),
        ])
        _notify_change()
        return

    payload = card_for(r)
    chans = r.get("channels")
    if (r.get("tier") or 2) == tier.CRITICAL:
        payload["urgency"] = "warn"
        payload["icon"] = "🚨"
        # critical reaches you wherever you are, not only where it fired
        chans = ["card", "system"] + [c for c in (r.get("channels") or [])
                                      if c not in ("card", "system")]
        try:
            import winsound
            winsound.MessageBeep(winsound.MB_ICONEXCLAMATION)
        except Exception:
            pass
    delivered = notifier.notify(payload, {
        "channels": chans,
        "event": "reminder.fired",
        "reminder": r,
        # the same sentence the decision log gets: on a phone there is no
        # screen context to reconstruct why this arrived
        "trigger": {"gate": gate, "why": why_now(r, snap, gate)},
    })
    why.note("reminder fired", r["text"], [
        ("why now", why_now(r, snap, gate)),
        ("set", time.strftime("%I:%M %p", time.localtime(r.get("createdAt") or time.time())).lstrip("0")
                + (f' from "{r["whenPhrase"]}"' if r.get("whenPhrase") else "")),
        ("delivered to", " + ".join(delivered or r.get("channels") or ["card"])),
        ("waiting on", "you — the card stays up until you answer it"),
    ])
    _notify_change()


def restore_fired() -> int:
    """A sticky card only survives as long as the process drawing it. Put
    unanswered reminders back up on start — local card only: the remote
    channels already delivered once."""
    n = 0
    for r in reminders.active():
        if r.get("state") == "fired":
            n += 1
            notifier.notify(card_for(r), {
                "channels": ["card"],
                "no_mirror": True,  # a redraw, not a new event
                "event": "reminder.restored",
                "reminder": r,
            })
    if n:
        eventlog.append({"event": "trigger.restored", "count": n})
    return n


def retier(r, snap) -> None:
    """Tier is a property of the reminder *right now*, so it is recomputed
    rather than remembered — inside two hard limits (see crw.tier)."""
    from_ = r.get("tier") or tier.UPCOMING
    to, reason = from_, None

    if r.get("dueAt"):
        # Proximity raises volume, but ambient promised to never interrupt,
        # and critical needs stated consequence, not a clock — nearness alone
        # caps out at upcoming.
        left = r["dueAt"] - int(time.time())
        if from_ != tier.AMBIENT and left <= 3600 and from_ > tier.UPCOMING:
            to, reason = tier.UPCOMING, "within the hour"
    elif (r.get("referent") and from_ == tier.AMBIENT
          and matcher.matches(r["referent"], snap)):
        # something you were only tracking is now in front of you
        to, reason = tier.INCONTEXT, "the thing it's about is on screen now"
    elif (from_ == tier.INCONTEXT and r.get("state") == "pending"
          and int(time.time()) - (r.get("createdAt") or int(time.time())) > 86400):
        # a context that hasn't come back in a day isn't going to interrupt well
        to, reason = tier.AMBIENT, "its context hasn't appeared in a day"

    if to != from_:
        r["tier"], r["tierWhy"] = to, reason
        r.setdefault("tierHistory", []).append(
            {"t": int(time.time()), "from": from_, "to": to, "why": reason})
        reminders.persist()
        eventlog.append({"event": "tier.moved", "id": r["id"],
                         "from": from_, "to": to, "why": reason})
        why.note("reminder got louder" if to < from_ else "reminder got quieter",
                 r["text"], [
                     ("moved", f'{tier.LABEL[from_]["name"]} → {tier.LABEL[to]["name"]}'),
                     ("because", reason),
                     ("now", tier.LABEL[to]["note"]),
                 ])
        _notify_change()


def check_due(snap) -> None:
    """Timed reminders bypass the FSM entirely: they fire on the clock,
    wherever the user is. Called every second for demo-friendly precision."""
    now = int(time.time())
    for r in reminders.active():
        if r.get("state") == "scheduled" and r.get("dueAt") and now >= r["dueAt"]:
            try:
                fire(r, snap or {}, "time")
            except Exception as e:
                eventlog.append({"event": "trigger.error", "id": r["id"],
                                 "error": str(e)})


def _step(r, snap) -> None:
    present = matcher.matches(r.get("referent"), snap)
    st = r.get("state")

    if st == "pending":
        if present:
            r["absent"] = 0
            reminders.set_state(r, "armed")

    elif st == "armed":
        if not present:
            r["absent"] = 1
            reminders.set_state(r, "cooldown")

    elif st == "cooldown":
        if present:
            # brief tab-away, not "done" — the debounce doing its job
            r["absent"] = 0
            reminders.set_state(r, "armed", "rearmed")
        else:
            r["absent"] = (r.get("absent") or 0) + 1
            if r["absent"] >= config.ABSENT_SAMPLES:
                r["readyAt"] = int(time.time())
                reminders.set_state(r, "ready")

    elif st == "ready":
        if present:
            # came back before we fired: definitely not done
            r["absent"] = 0
            reminders.set_state(r, "armed", "returned")
        else:
            gate = None
            if tier.bypasses_seam_gate(r.get("tier")):
                # the one tier allowed to interrupt mid-focus — the stated
                # exception, not an oversight
                gate = "critical"
            elif snap.get("reason") in ("app-switch", "wake"):
                gate = snap["reason"]
            elif int(time.time()) - (r.get("readyAt") or 0) >= config.MAX_READY_WAIT:
                gate = "max-wait"
            if gate:
                fire(r, snap, gate)


def tick(snap) -> None:
    """One observer sample through every active reminder. Exported for
    scripted tests: feed synthetic snapshots without waiting on wall-clock."""
    for r in reminders.active():
        try:
            retier(r, snap)
        except Exception as e:
            eventlog.append({"event": "trigger.error", "id": r.get("id"),
                             "error": "retier: " + str(e)})
        before = r.get("state")
        try:
            _step(r, snap)
        except Exception as e:
            eventlog.append({"event": "trigger.error", "id": r.get("id"),
                             "error": str(e)})
        if r.get("state") != before:
            _notify_change()
